diff --git js/defs.bzl js/defs.bzl
index fd8d5615..1ffabfea 100644
--- js/defs.bzl
+++ js/defs.bzl
@@ -9,6 +9,7 @@ load(
 load(
     "//js/private:js_image_layer.bzl",
     _js_image_layer = "js_image_layer",
+    _DEFAULT_LAYER_GROUPS = "DEFAULT_LAYER_GROUPS"
 )
 load(
     "//js/private:js_info_files.bzl",
@@ -52,3 +53,4 @@ js_run_devserver = _js_run_devserver
 js_run_binary = _js_run_binary
 js_info_files = _js_info_files
 js_image_layer = _js_image_layer
+DEFAULT_LAYER_GROUPS = _DEFAULT_LAYER_GROUPS
diff --git js/private/js_image_layer.bzl js/private/js_image_layer.bzl
index 6e28a6fd..551247bb 100644
--- js/private/js_image_layer.bzl
+++ js/private/js_image_layer.bzl
@@ -571,3 +571,5 @@ js_image_layer = rule(
         "@rules_nodejs//nodejs:toolchain_type",
     ],
 )
+
+DEFAULT_LAYER_GROUPS = _DEFAULT_LAYER_GROUPS
diff --git npm/private/npm_import.bzl npm/private/npm_import.bzl
index 0cf4f2f..6d7e230 100644
--- npm/private/npm_import.bzl
+++ npm/private/npm_import.bzl
@@ -161,7 +161,7 @@ def npm_imported_package_store_internal(
         srcs = [":{}".format(store_target_name)],
         output_group = utils.package_directory_output_group,
         visibility = ["//visibility:public"],
-        tags = ["manual"],
+        tags = ["manual", "no-lint"],
     )
 
     if has_lifecycle_build_target:
@@ -268,7 +268,7 @@ def npm_link_imported_package_store_internal(
         srcs = [":" + target_name],
         output_group = utils.package_directory_output_group,
         visibility = link_visibility,
-        tags = ["manual"],
+        tags = ["manual", "no-lint"],
     )
 
 _LINK_JS_PACKAGE_LINK_IMPORTED_STORE_TMPL = """\
diff --git npm/private/npm_link_package.bzl npm/private/npm_link_package.bzl
index be0f25f..82a9e9e 100644
--- npm/private/npm_link_package.bzl
+++ npm/private/npm_link_package.bzl
@@ -44,6 +44,9 @@ def npm_link_package(
     if auto_manual and "manual" not in tags:
         tags.append("manual")
 
+    if "no-lint" not in tags:
+        tags.append("no-lint")
+
     # link the package store when linking at the root
     npm_package_store(
         name = store_target_name,
diff --git npm/private/npm_link_package_store.bzl npm/private/npm_link_package_store.bzl
index 82bf82d..e32c3ef 100644
--- npm/private/npm_link_package_store.bzl
+++ npm/private/npm_link_package_store.bzl
@@ -163,5 +163,5 @@ def npm_local_link_package_store_internal(name, src, package = None, link_visibi
         srcs = [":" + name],
         output_group = utils.package_directory_output_group,
         visibility = link_visibility,
-        tags = ["manual"],
+        tags = ["manual", "no-lint"],
     )
diff --git npm/private/npm_translate_lock_generate.bzl npm/private/npm_translate_lock_generate.bzl
index 1dd02af..49af2bf 100644
--- npm/private/npm_translate_lock_generate.bzl
+++ npm/private/npm_translate_lock_generate.bzl
@@ -23,7 +23,7 @@ _FP_STORE_TMPL = \
             version = "{version}",
             deps = {deps},
             visibility = ["//visibility:public"],
-            tags = ["manual"],
+            tags = ["manual", "no-lint"],
         )"""
 
 _FP_DIRECT_TMPL = \
