diff --git MODULE.bazel MODULE.bazel
new file mode 100644
index 00000000..0250a2fa
--- /dev/null
+++ MODULE.bazel
@@ -0,0 +1,41 @@
+module(name = "build_stack_rules_proto")
+
+bazel_dep(name = "aspect_rules_js")
+bazel_dep(name = "aspect_rules_ts")
+bazel_dep(name = "gazelle", version = "0.31.0", repo_name = "bazel_gazelle")
+bazel_dep(name = "grpc", version = "1.48.1", repo_name = "com_github_grpc_grpc")
+bazel_dep(name = "protobuf", repo_name = "com_google_protobuf")
+bazel_dep(name = "rules_cc")
+bazel_dep(name = "rules_go", repo_name = "io_bazel_rules_go")
+bazel_dep(name = "rules_proto", version = "5.3.0-21.7")
+bazel_dep(name = "rules_python")
+
+npm = use_extension(
+    "@aspect_rules_js//npm:extensions.bzl",
+    "npm",
+)
+npm.npm_translate_lock(
+    name = "npm_ts_proto",
+    generate_bzl_library_targets = True,
+    link_workspace = "build_stack_rules_proto",
+    npmrc = "//:.npmrc",
+    pnpm_lock = "//:pnpm-lock.yaml",
+    verify_node_modules_ignored = "//:.bazelignore",
+)
+use_repo(npm, "npm_ts_proto")
+
+go_deps = use_extension("@bazel_gazelle//:extensions.bzl", "go_deps")
+go_deps.from_file(go_mod = "//:go.mod")
+use_repo(
+    go_deps,
+    "com_github_bazelbuild_buildtools",
+    "com_github_bmatcuk_doublestar",
+    "com_github_emicklei_proto",
+    "com_github_golang_glog",
+    "com_github_grpc_ecosystem_grpc_gateway_v2",
+    "in_gopkg_yaml_v3",
+    "net_starlark_go",
+    "org_golang_google_genproto_googleapis_api",
+    "org_golang_google_grpc_cmd_protoc_gen_go_grpc",
+    "org_golang_google_protobuf",
+)
diff --git deps/BUILD.bazel deps/BUILD.bazel
index 01c28bc4..a59f6469 100644
--- deps/BUILD.bazel
+++ deps/BUILD.bazel
@@ -103,19 +103,19 @@ proto_dependency(
     visibility = ["//visibility:public"],
 )
 
-# Release: v0.31.0
+# Release: v0.35.0
 # TargetCommitish: master
-# Date: 2023-05-27 18:23:20 +0000 UTC
-# URL: https://github.com/bazelbuild/bazel-gazelle/releases/tag/v0.31.0
-# Size: 1870962 (1.9 MB)
+# Date: 2023-12-21 18:18:45 +0000 UTC
+# URL: https://github.com/bazelbuild/bazel-gazelle/releases/tag/v0.35.0
+# Size: 1780790 (1.7 MB)
 proto_dependency(
     name = "bazel_gazelle",
     patch_args = ["-p1"],
     patches = ["@build_stack_rules_proto//third_party:bazel-gazelle-revert-1152.patch"],
     repository_rule = "http_archive",
-    sha256 = "bc9a8c259ad2eb54dd89404979c097451df8f2dc64852c9f38d2b7a248f84f32",
-    strip_prefix = "bazel-gazelle-0.31.0",
-    urls = ["https://github.com/bazelbuild/bazel-gazelle/archive/v0.31.0.tar.gz"],
+    sha256 = "a0ee1d304f7caa46680ba06bdef0e5d9ec8815f6e01ec29398efd13256598c3f",
+    strip_prefix = "bazel-gazelle-0.35.0",
+    urls = ["https://github.com/bazelbuild/bazel-gazelle/archive/v0.35.0.tar.gz"],
     deps = [":io_bazel_rules_go"],
 )
 
diff --git deps/core_deps.bzl deps/core_deps.bzl
index 70b2d8c5..558b9444 100644
--- deps/core_deps.bzl
+++ deps/core_deps.bzl
@@ -30,10 +30,10 @@ def bazel_gazelle():
     _maybe(
         http_archive,
         name = "bazel_gazelle",
-        sha256 = "bc9a8c259ad2eb54dd89404979c097451df8f2dc64852c9f38d2b7a248f84f32",
-        strip_prefix = "bazel-gazelle-0.31.0",
+        sha256 = "a0ee1d304f7caa46680ba06bdef0e5d9ec8815f6e01ec29398efd13256598c3f",
+        strip_prefix = "bazel-gazelle-0.35.0",
         urls = [
-            "https://github.com/bazelbuild/bazel-gazelle/archive/v0.31.0.tar.gz",
+            "https://github.com/bazelbuild/bazel-gazelle/archive/v0.35.0.tar.gz",
         ],
         patches = [
             "@build_stack_rules_proto//third_party:bazel-gazelle-revert-1152.patch",
diff --git go.mod go.mod
index a6f8c4b6..ecd76157 100644
--- go.mod
+++ go.mod
@@ -1,6 +1,6 @@
 module github.com/stackb/rules_proto
 
-go 1.15
+go 1.17
 
 require (
 	github.com/bazelbuild/bazel-gazelle v0.31.0
@@ -9,8 +9,22 @@ require (
 	github.com/bmatcuk/doublestar v1.2.2
 	github.com/emicklei/proto v1.13.0
 	github.com/google/go-cmp v0.5.9
+	github.com/grpc-ecosystem/grpc-gateway/v2 v2.16.0
 	github.com/pmezard/go-difflib v1.0.0
 	github.com/stretchr/testify v1.7.0
 	go.starlark.net v0.0.0-20220328144851-d1966c6b9fcd
-	gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b
+	google.golang.org/grpc/cmd/protoc-gen-go-grpc v1.3.0
+	gopkg.in/yaml.v3 v3.0.1
+	google.golang.org/genproto v0.0.0-20230530153820-e85fd2cbaebc
 )
+
+require (
+	github.com/bmatcuk/doublestar/v4 v4.6.0 // indirect
+	github.com/davecgh/go-spew v1.1.0 // indirect
+	github.com/golang/glog v1.1.1 // indirect
+	golang.org/x/mod v0.10.0 // indirect
+	golang.org/x/sync v0.2.0 // indirect
+	golang.org/x/sys v0.8.0 // indirect
+	golang.org/x/tools v0.9.1 // indirect
+	google.golang.org/protobuf v1.30.0 // indirect
+ )
diff --git go.sum go.sum
index c720968b..96780d19 100644
--- go.sum
+++ go.sum
@@ -41,6 +41,8 @@ github.com/envoyproxy/protoc-gen-validate v0.1.0/go.mod h1:iSmxcyjqTsJpI2R4NaDN7
 github.com/fsnotify/fsnotify v1.6.0/go.mod h1:sl3t1tCWJFWoRz9R8WJCbQihKKwmorjAbSClcnxKAGw=
 github.com/ghodss/yaml v1.0.0/go.mod h1:4dBDuWmgqj2HViK6kFavaiC9ZROes6MMH2rRYeMEF04=
 github.com/golang/glog v0.0.0-20160126235308-23def4e6c14b/go.mod h1:SBH7ygxi8pfUlaOkMMuAQtPIUF8ecWP5IEl/CR7VP2Q=
+github.com/golang/glog v1.1.1 h1:jxpi2eWoU84wbX9iIEyAeeoac3FLuifZpY9tcNUD9kw=
+github.com/golang/glog v1.1.1/go.mod h1:zR+okUeTbrL6EL3xHUDxZuEtGv04p5shwip1+mL/rLQ=
 github.com/golang/mock v1.1.1/go.mod h1:oTYuIxOrZwtPieC+H1uAHpcLFnEyAGVDL/k47Jfbm0A=
 github.com/golang/mock v1.6.0/go.mod h1:p6yTPP+5HYm5mzsMV8JkE6ZKdX+/wYM6Hr+LicevLPs=
 github.com/golang/protobuf v1.2.0/go.mod h1:6lQm79b+lXiMfvg/cZm0SGofjICqVBUtrP5yJMmIC1U=
@@ -68,6 +70,8 @@ github.com/google/go-cmp v0.5.9 h1:O2Tfq5qg4qc4AmwVlvv0oLiVAGB7enBSJ2x2DqQFi38=
 github.com/google/go-cmp v0.5.9/go.mod h1:17dUlkBOakJ0+DkrSSNjCkIjxS6bF9zb3elmeNGIjoY=
 github.com/google/uuid v1.1.2/go.mod h1:TIyPZe4MgqvfeYDBFedMoGGpEw/LqOeaOT+nhxU+yHo=
 github.com/grpc-ecosystem/grpc-gateway v1.16.0/go.mod h1:BDjrQk3hbvj6Nolgz8mAMFbcEtjT1g+wF4CSlocrBnw=
+github.com/grpc-ecosystem/grpc-gateway/v2 v2.16.0 h1:YBftPWNWd4WwGqtY2yeZL2ef8rHAxPBD8KFhJpmcqms=
+github.com/grpc-ecosystem/grpc-gateway/v2 v2.16.0/go.mod h1:YN5jB8ie0yfIUg6VvR9Kz84aCaG7AsGZnLjhHbUqwPg=
 github.com/pmezard/go-difflib v1.0.0 h1:4DBwDE0NGyQoBHbLQYPwSUPoCMWR5BEzIk/f1lZbAQM=
 github.com/pmezard/go-difflib v1.0.0/go.mod h1:iKH77koFhYxTK1pcRnkKkqfTogsbg7gZNVY4sRDYZ/4=
 github.com/prometheus/client_model v0.0.0-20190812154241-14fe0d1b01d4/go.mod h1:xMI15A0UPsDsEKsMN9yxemIoYk6Tm2C1GtYGdfGttqA=
@@ -166,6 +170,8 @@ google.golang.org/genproto v0.0.0-20180817151627-c66870c02cf8/go.mod h1:JiN7NxoA
 google.golang.org/genproto v0.0.0-20190819201941-24fa4b261c55/go.mod h1:DMBHOl98Agz4BDEuKkezgsaosCRResVns1a3J2ZsMNc=
 google.golang.org/genproto v0.0.0-20200513103714-09dca8ec2884/go.mod h1:55QSHmfGQM9UVYDPBsyGGes0y52j32PQ3BqQfXhyH3c=
 google.golang.org/genproto v0.0.0-20200526211855-cb27e3aa2013/go.mod h1:NbSheEEYHJ7i3ixzK3sjbqSGDJWnxyFXZblF3eUsNvo=
+google.golang.org/genproto v0.0.0-20230530153820-e85fd2cbaebc h1:8DyZCyvI8mE1IdLy/60bS+52xfymkE72wv1asokgtao=
+google.golang.org/genproto v0.0.0-20230530153820-e85fd2cbaebc/go.mod h1:xZnkP7mREFX5MORlOPEzLMr+90PPZQ2QWzrVTWfAq64=
 google.golang.org/grpc v1.19.0/go.mod h1:mqu4LbDTu4XGKhr4mRzUsmM4RtVoemTSY81AxZiDr8c=
 google.golang.org/grpc v1.23.0/go.mod h1:Y5yQAOtifL1yxbo5wqy6BxZv8vAUGQwXBOALyacEbxg=
 google.golang.org/grpc v1.25.1/go.mod h1:c3i+UQWmh7LiEpx4sFZnkU36qjEYZ0imhYfXVyQciAY=
@@ -173,6 +179,8 @@ google.golang.org/grpc v1.27.0/go.mod h1:qbnxyOmOxrQa7FizSgH+ReBfzJrCY1pSN7KXBS8
 google.golang.org/grpc v1.33.1/go.mod h1:fr5YgcSWrqhRRxogOsw7RzIpsmvOZ6IcH4kBYTpR3n0=
 google.golang.org/grpc v1.36.0/go.mod h1:qjiiYl8FncCW8feJPdyg3v6XW24KsRHe+dy9BAGRRjU=
 google.golang.org/grpc v1.50.0/go.mod h1:ZgQEeidpAuNRZ8iRrlBKXZQP1ghovWIVhdJRyCDK+GI=
+google.golang.org/grpc/cmd/protoc-gen-go-grpc v1.3.0 h1:rNBFJjBCOgVr9pWD7rs/knKL4FRTKgpZmsRfV214zcA=
+google.golang.org/grpc/cmd/protoc-gen-go-grpc v1.3.0/go.mod h1:Dk1tviKTvMCz5tvh7t+fh94dhmQVHuCt2OzJB3CTW9Y=
 google.golang.org/protobuf v0.0.0-20200109180630-ec00e32a8dfd/go.mod h1:DFci5gLYBciE7Vtevhsrf46CRTquxDuWsQurQQe4oz8=
 google.golang.org/protobuf v0.0.0-20200221191635-4d8936d0db64/go.mod h1:kwYJMbMJ01Woi6D6+Kah6886xMZcty6N08ah7+eCXa0=
 google.golang.org/protobuf v0.0.0-20200228230310-ab0ca4ff8a60/go.mod h1:cfTl7dwQJ+fmap5saPgwCLgHXTUD7jkjRqWcaiX5VyM=
@@ -186,6 +194,8 @@ google.golang.org/protobuf v1.26.0-rc.1/go.mod h1:jlhhOSvTdKEhbULTjvd4ARK9grFBp0
 google.golang.org/protobuf v1.26.0/go.mod h1:9q0QmTI4eRPtz6boOQmLYwt+qCgq0jsYwAQnmE0givc=
 google.golang.org/protobuf v1.27.1/go.mod h1:9q0QmTI4eRPtz6boOQmLYwt+qCgq0jsYwAQnmE0givc=
 google.golang.org/protobuf v1.28.0/go.mod h1:HV8QOd/L58Z+nl8r43ehVNZIU/HEI6OcFqwMG9pJV4I=
+google.golang.org/protobuf v1.30.0 h1:kPPoIgf3TsEvrm0PFe15JQ+570QVxYzEvvHqChK+cng=
+google.golang.org/protobuf v1.30.0/go.mod h1:HV8QOd/L58Z+nl8r43ehVNZIU/HEI6OcFqwMG9pJV4I=
 gopkg.in/check.v1 v0.0.0-20161208181325-20d25e280405 h1:yhCVgyC4o1eVCa2tZl7eS0r+SDo693bJlVdllGtEeKM=
 gopkg.in/check.v1 v0.0.0-20161208181325-20d25e280405/go.mod h1:Co6ibVJAznAaIkqp8huTwlJQCZ016jof/cbN4VW5Yz0=
 gopkg.in/yaml.v2 v2.2.2/go.mod h1:hI93XBmqTisBFMUTm0b8Fm+jr3Dg1NNxqwp+5A1VGuI=
@@ -193,5 +203,7 @@ gopkg.in/yaml.v2 v2.2.3/go.mod h1:hI93XBmqTisBFMUTm0b8Fm+jr3Dg1NNxqwp+5A1VGuI=
 gopkg.in/yaml.v3 v3.0.0-20200313102051-9f266ea9e77c/go.mod h1:K4uyk7z7BCEPqu6E+C64Yfv1cQ7kz7rIZviUmN+EgEM=
 gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b h1:h8qDotaEPuJATrMmW04NCwg7v22aHH28wwpauUhK9Oo=
 gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b/go.mod h1:K4uyk7z7BCEPqu6E+C64Yfv1cQ7kz7rIZviUmN+EgEM=
+gopkg.in/yaml.v3 v3.0.1 h1:fxVm/GzAzEWqLHuvctI91KS9hhNmmWOoWu0XTYJS7CA=
+gopkg.in/yaml.v3 v3.0.1/go.mod h1:K4uyk7z7BCEPqu6E+C64Yfv1cQ7kz7rIZviUmN+EgEM=
 honnef.co/go/tools v0.0.0-20190102054323-c2f93a96b099/go.mod h1:rf3lG4BRIbNafJWhAfAdb/ePZxsR/4RtNHQocxwk9r4=
 honnef.co/go/tools v0.0.0-20190523083050-ea95bdfd59fc/go.mod h1:rf3lG4BRIbNafJWhAfAdb/ePZxsR/4RtNHQocxwk9r4=
diff --git pkg/language/protobuf/config.go pkg/language/protobuf/config.go
index 4b65f556..e1644bbf 100644
--- pkg/language/protobuf/config.go
+++ pkg/language/protobuf/config.go
@@ -130,18 +130,18 @@ func (pl *protobufLang) getOrCreatePackageConfig(config *config.Config) *protoc.
 
 func registerWellKnownProtos(resolver protoc.ImportResolver) {
 	for k, v := range map[string]label.Label{
-		"google/protobuf/any.proto":             label.New("com_google_protobuf", "", "any_proto"),
-		"google/protobuf/api.proto":             label.New("com_google_protobuf", "", "api_proto"),
-		"google/protobuf/compiler/plugin.proto": label.New("com_google_protobuf", "", "compiler_plugin_proto"),
-		"google/protobuf/descriptor.proto":      label.New("com_google_protobuf", "", "descriptor_proto"),
-		"google/protobuf/duration.proto":        label.New("com_google_protobuf", "", "duration_proto"),
-		"google/protobuf/empty.proto":           label.New("com_google_protobuf", "", "empty_proto"),
-		"google/protobuf/field_mask.proto":      label.New("com_google_protobuf", "", "field_mask_proto"),
-		"google/protobuf/source_context.proto":  label.New("com_google_protobuf", "", "source_context_proto"),
-		"google/protobuf/struct.proto":          label.New("com_google_protobuf", "", "struct_proto"),
-		"google/protobuf/timestamp.proto":       label.New("com_google_protobuf", "", "timestamp_proto"),
-		"google/protobuf/type.proto":            label.New("com_google_protobuf", "", "type_proto"),
-		"google/protobuf/wrappers.proto":        label.New("com_google_protobuf", "", "wrappers_proto"),
+		"google/protobuf/any.proto":             label.New("protobuf", "", "any_proto"),
+		"google/protobuf/api.proto":             label.New("protobuf", "", "api_proto"),
+		"google/protobuf/compiler/plugin.proto": label.New("protobuf", "", "compiler_plugin_proto"),
+		"google/protobuf/descriptor.proto":      label.New("protobuf", "", "descriptor_proto"),
+		"google/protobuf/duration.proto":        label.New("protobuf", "", "duration_proto"),
+		"google/protobuf/empty.proto":           label.New("protobuf", "", "empty_proto"),
+		"google/protobuf/field_mask.proto":      label.New("protobuf", "", "field_mask_proto"),
+		"google/protobuf/source_context.proto":  label.New("protobuf", "", "source_context_proto"),
+		"google/protobuf/struct.proto":          label.New("protobuf", "", "struct_proto"),
+		"google/protobuf/timestamp.proto":       label.New("protobuf", "", "timestamp_proto"),
+		"google/protobuf/type.proto":            label.New("protobuf", "", "type_proto"),
+		"google/protobuf/wrappers.proto":        label.New("protobuf", "", "wrappers_proto"),
 	} {
 		resolver.Provide("proto", "proto", k, v)
 	}
diff --git pkg/plugin/grpcecosystem/grpcgateway/BUILD.bazel pkg/plugin/grpcecosystem/grpcgateway/BUILD.bazel
index 72fa9c4f..227b20e0 100644
--- pkg/plugin/grpcecosystem/grpcgateway/BUILD.bazel
+++ pkg/plugin/grpcecosystem/grpcgateway/BUILD.bazel
@@ -2,7 +2,10 @@ load("@io_bazel_rules_go//go:def.bzl", "go_library", "go_test")
 
 go_library(
     name = "grpcgateway",
-    srcs = ["protoc-gen-grpc-gateway.go"],
+    srcs = [
+        "protoc-gen-grpc-gateway.go",
+        "protoc-gen-grpc-openapiv2.go",
+    ],
     importpath = "github.com/stackb/rules_proto/pkg/plugin/grpcecosystem/grpcgateway",
     visibility = ["//visibility:public"],
     deps = [
diff --git pkg/plugin/grpcecosystem/grpcgateway/protoc-gen-grpc-openapiv2.go pkg/plugin/grpcecosystem/grpcgateway/protoc-gen-grpc-openapiv2.go
new file mode 100644
index 00000000..4a8524e2
--- /dev/null
+++ pkg/plugin/grpcecosystem/grpcgateway/protoc-gen-grpc-openapiv2.go
@@ -0,0 +1,51 @@
+package grpcgateway
+
+import (
+	"path"
+
+	"github.com/bazelbuild/bazel-gazelle/label"
+	"github.com/stackb/rules_proto/pkg/protoc"
+)
+
+func init() {
+	protoc.Plugins().MustRegisterPlugin(&protocGenGrpcOpenapiv2Plugin{})
+}
+
+// protocGenGrpcOpenapiv2Plugin implements Plugin for protoc-gen-grpc-openapiv2.
+type protocGenGrpcOpenapiv2Plugin struct{}
+
+// Name implements part of the Plugin interface.
+func (p *protocGenGrpcOpenapiv2Plugin) Name() string {
+	return "grpc-ecosystem:grpc-gateway:protoc-gen-grpc-openapiv2"
+}
+
+// Configure implements part of the Plugin interface.
+func (p *protocGenGrpcOpenapiv2Plugin) Configure(ctx *protoc.PluginContext) *protoc.PluginConfiguration {
+	if !p.shouldApply(ctx.ProtoLibrary) {
+		return nil
+	}
+	return &protoc.PluginConfiguration{
+		Label:   label.New("build_stack_rules_proto", "plugin/grpc-ecosystem/grpc-gateway", "protoc-gen-grpc-openapiv2"),
+		Outputs: p.outputs(ctx.Rel, ctx.ProtoLibrary),
+		Options: ctx.PluginConfig.GetOptions(),
+	}
+}
+
+func (p *protocGenGrpcOpenapiv2Plugin) shouldApply(lib protoc.ProtoLibrary) bool {
+	return protoc.HasServices(lib.Files()...)
+}
+
+func (p *protocGenGrpcOpenapiv2Plugin) outputs(rel string, lib protoc.ProtoLibrary) []string {
+	srcs := make([]string, 0)
+	for _, f := range lib.Files() {
+		if !f.HasServices() {
+			continue
+		}
+		base := f.Name
+		if rel != "" {
+			base = path.Join(rel, base)
+		}
+		srcs = append(srcs, base+".swagger.json")
+	}
+	return srcs
+}
diff --git pkg/protoc/language_config.go pkg/protoc/language_config.go
index c016089e..997e11da 100644
--- pkg/protoc/language_config.go
+++ pkg/protoc/language_config.go
@@ -2,6 +2,7 @@ package protoc
 
 import (
 	"fmt"
+	"slices"
 	"sort"
 	"strconv"
 )
@@ -9,21 +10,23 @@ import (
 // LanguageConfig carries a set of configured Plugins and Rules that will
 // contribute to a protoc invocation.
 type LanguageConfig struct {
-	Name    string
-	Protoc  string
-	Enabled bool
-	Plugins map[string]bool
-	Rules   map[string]bool
+	Name         string
+	Protoc       string
+	Enabled      bool
+	Plugins      map[string]bool
+	PluginsOrder []string
+	Rules        map[string]bool
 }
 
 // newLanguageConfig constructs a new language configuration having the
 // given name and list of default rule configurations.
 func newLanguageConfig(name string, rules ...*LanguageRuleConfig) *LanguageConfig {
 	c := &LanguageConfig{
-		Name:    name,
-		Rules:   make(map[string]bool),
-		Plugins: make(map[string]bool),
-		Enabled: true,
+		Name:         name,
+		Rules:        make(map[string]bool),
+		Plugins:      make(map[string]bool),
+		PluginsOrder: make([]string, 0),
+		Enabled:      true,
 	}
 	for _, rule := range rules {
 		c.Rules[rule.Name] = true
@@ -51,6 +54,9 @@ func (c *LanguageConfig) clone() *LanguageConfig {
 	for k, v := range c.Plugins {
 		clone.Plugins[k] = v
 	}
+	for _, plugin := range c.PluginsOrder {
+		clone.PluginsOrder = append(clone.PluginsOrder, plugin)
+	}
 	for k, v := range c.Rules {
 		clone.Rules[k] = v
 	}
@@ -69,6 +75,9 @@ func (c *LanguageConfig) parseDirective(cfg *PackageConfig, d, param, value stri
 		c.Enabled = enabled
 	case "plugin":
 		c.Plugins[value] = intent.Want
+		if !slices.Contains(c.PluginsOrder, value) {
+			c.PluginsOrder = append(c.PluginsOrder, value)
+		}
 	case "protoc":
 		c.Protoc = value
 	case "rule":
@@ -80,13 +89,16 @@ func (c *LanguageConfig) parseDirective(cfg *PackageConfig, d, param, value stri
 	return nil
 }
 
-// fromYAML loads configuration from the yaml rule confug.
+// fromYAML loads configuration from the yaml rule config.
 func (c *LanguageConfig) fromYAML(y *YLanguage) error {
 	if c.Name != y.Name {
 		return fmt.Errorf("yaml language mismatch: want %q got %q", c.Name, y.Name)
 	}
 	for _, plugin := range y.Plugin {
 		c.Plugins[plugin] = true
+		if !slices.Contains(c.PluginsOrder, plugin) {
+			c.PluginsOrder = append(c.PluginsOrder, plugin)
+		}
 	}
 	for _, rule := range y.Rule {
 		c.Rules[rule] = true
diff --git pkg/protoc/package.go pkg/protoc/package.go
index 8465db82..2b64f09e 100644
--- pkg/protoc/package.go
+++ pkg/protoc/package.go
@@ -62,56 +62,56 @@ func (s *Package) libraryRules(p *LanguageConfig, lib ProtoLibrary) []RuleProvid
 	// list of plugin configurations that apply to this proto_library
 	configs := make([]*PluginConfiguration, 0)
 
-	for name, want := range p.Plugins {
-		if !want {
-			continue
-		}
-		plugin, ok := s.cfg.plugins[name]
-		if !ok {
-			log.Fatalf("plugin not configured: %q", name)
-		}
-		if !plugin.Enabled {
-			continue
-		}
+	for _, name := range p.PluginsOrder {
+		want := p.Plugins[name]
+		if want {
+			plugin, ok := s.cfg.plugins[name]
+			if !ok {
+				log.Fatalf("plugin not configured: %q", name)
+			}
+			if !plugin.Enabled {
+				continue
+			}
 
-		ctx := &PluginContext{
-			Rel:           s.rel,
-			ProtoLibrary:  lib,
-			PackageConfig: *s.cfg,
-			PluginConfig:  *plugin,
-		}
+			ctx := &PluginContext{
+				Rel:           s.rel,
+				ProtoLibrary:  lib,
+				PackageConfig: *s.cfg,
+				PluginConfig:  *plugin,
+			}
 
-		if plugin.Implementation == "" {
-			plugin.Implementation = plugin.Name
-		}
-		impl, err := globalRegistry.LookupPlugin(plugin.Implementation)
+			if plugin.Implementation == "" {
+				plugin.Implementation = plugin.Name
+			}
+			impl, err := globalRegistry.LookupPlugin(plugin.Implementation)
 
-		if err == ErrUnknownPlugin {
-			log.Fatalf(
-				"%s: plugin not registered: %q (available: %v) [%+v]",
-				s.rel,
-				plugin.Implementation,
-				globalRegistry.PluginNames(),
-				plugin,
-			)
-		}
-		ctx.Plugin = impl
+			if err == ErrUnknownPlugin {
+				log.Fatalf(
+					"%s: plugin not registered: %q (available: %v) [%+v]",
+					s.rel,
+					plugin.Implementation,
+					globalRegistry.PluginNames(),
+					plugin,
+				)
+			}
+			ctx.Plugin = impl
 
-		// Delegate to the implementation for configuration
-		config := impl.Configure(ctx)
-		if config == nil {
-			continue
-		}
-		config.Plugin = impl
-		config.Config = plugin.clone()
-		config.Options = DeduplicateAndSort(config.Options)
+			// Delegate to the implementation for configuration
+			config := impl.Configure(ctx)
+			if config == nil {
+				continue
+			}
+			config.Plugin = impl
+			config.Config = plugin.clone()
+			config.Options = DeduplicateAndSort(config.Options)
 
-		// plugin.Label overrides the default value from the implementation
-		if plugin.Label.Name != "" {
-			config.Label = plugin.Label
-		}
+			// plugin.Label overrides the default value from the implementation
+			if plugin.Label.Name != "" {
+				config.Label = plugin.Label
+			}
 
-		configs = append(configs, config)
+			configs = append(configs, config)
+		}
 	}
 
 	if len(configs) == 0 {
diff --git pkg/protoc/plugin_configuration.go pkg/protoc/plugin_configuration.go
index 8c3d7d29..03365de4 100644
--- pkg/protoc/plugin_configuration.go
+++ pkg/protoc/plugin_configuration.go
@@ -35,7 +35,6 @@ func GetPluginLabels(plugins []*PluginConfiguration) []string {
 	for i, plugin := range plugins {
 		labels[i] = plugin.Label.String()
 	}
-	sort.Strings(labels)
 	return labels
 }
 
diff --git pkg/protoc/resolver.go pkg/protoc/resolver.go
index c2371006..0ce5a483 100644
--- pkg/protoc/resolver.go
+++ pkg/protoc/resolver.go
@@ -196,12 +196,8 @@ func (r *resolver) Resolve(lang, impLang, imp string) []resolve.FindResult {
 	}
 	if got, ok := known[imp]; ok {
 		res := make([]resolve.FindResult, len(got))
-		// use last-wins semantics by providing FindResults in reverse to how
-		// they were provided.
-		var index int
-		for i := len(got) - 1; i >= 0; i-- {
-			res[index] = resolve.FindResult{Label: got[i]}
-			index++
+		for i, l := range got {
+			res[i] = resolve.FindResult{Label: l}
 		}
 		return res
 	}
@@ -270,20 +266,16 @@ func (r *resolver) Install(c *config.Config) {
 	// The resolve config has already processed resolve directives, and there's
 	// no public API. Take somewhat extreme measures to augment it's internal
 	// override list via unsafe memory reallocation.
-	overrides := make([]overrideSpec, 0)
+	overrides := make(overrideSpec, 0)
 
 	for key, known := range r.known {
 		lang, impLang := keyLang(key)
 		for imp, lbls := range known {
 			for _, lbl := range lbls {
-				overrides = append(overrides, overrideSpec{
-					imp: resolve.ImportSpec{
-						Lang: impLang,
-						Imp:  imp,
-					},
+				overrides[overrideKey{
+					imp:  resolve.ImportSpec{Lang: impLang, Imp: imp},
 					lang: lang,
-					dep:  lbl,
-				})
+				}] = lbl
 			}
 		}
 	}
@@ -333,31 +325,36 @@ func getResolveConfig(c *config.Config) interface{} {
 
 // rewriteResolveConfigOverrides reads the existing private attribute and
 // appends more overrides.
-func rewriteResolveConfigOverrides(rc interface{}, more []overrideSpec) {
+func rewriteResolveConfigOverrides(rc interface{}, more overrideSpec) {
 	rcv := reflect.ValueOf(rc).Elem()
 	val := reflect.Indirect(rcv)
 	member := val.FieldByName("overrides")
 	ptrToOverrides := unsafe.Pointer(member.UnsafeAddr())
-	overrides := (*[]overrideSpec)(ptrToOverrides)
+	overrides := (*overrideSpec)(ptrToOverrides)
 
 	// create new array: FindRuleWithOverride searches last entries first, so
 	// respect the users own resolve directives by putting them last
-	newOverrides := make([]overrideSpec, 0)
-	newOverrides = append(newOverrides, more...)
-	newOverrides = append(newOverrides, *overrides...)
+	newOverrides := make(overrideSpec, 0)
+	for k, v := range more {
+		newOverrides[k] = v
+	}
+	for k, v := range *overrides {
+		newOverrides[k] = v
+	}
 
 	// reassign memory value
 	*overrides = newOverrides
 }
 
+type overrideKey struct {
+	imp  resolve.ImportSpec
+	lang string
+}
+
 // overrideSpec is a copy of the same private type in resolve/config.go.  It must be
 // kept in sync with the original to avoid discrepancy with the expected memory
 // layout.
-type overrideSpec struct {
-	imp  resolve.ImportSpec
-	lang string
-	dep  label.Label
-}
+type overrideSpec map[overrideKey]label.Label
 
 func langKey(lang, impLang string) string {
 	return lang + " " + impLang
diff --git pkg/rule/rules_python/py_library.go pkg/rule/rules_python/py_library.go
index 4d9f5b25..b42fdbe2 100644
--- pkg/rule/rules_python/py_library.go
+++ pkg/rule/rules_python/py_library.go
@@ -14,6 +14,7 @@ import (
 var pyLibraryKindInfo = rule.KindInfo{
 	MergeableAttrs: map[string]bool{
 		"srcs":       true,
+		"data":       true,
 		"deps":       true,
 		"visibility": true,
 		"imports":    true,
@@ -90,6 +91,15 @@ func (s *PyLibrary) Rule(otherGen ...*rule.Rule) *rule.Rule {
 	newRule := rule.NewRule(s.Kind(), s.Name())
 
 	newRule.SetAttr("srcs", s.Srcs())
+	var pyi_files []string
+	for _, src := range s.Srcs() {
+		if !strings.HasSuffix(src, "grpc.py") {
+			pyi_files = append(pyi_files, src+"i")
+		}
+	}
+	if len(pyi_files) > 0 {
+		newRule.SetAttr("data", pyi_files)
+	}
 
 	deps := s.Deps()
 	if len(deps) > 0 {
diff --git plugin/grpc-ecosystem/grpc-gateway/BUILD.bazel plugin/grpc-ecosystem/grpc-gateway/BUILD.bazel
index 98943320..047e27da 100644
--- plugin/grpc-ecosystem/grpc-gateway/BUILD.bazel
+++ plugin/grpc-ecosystem/grpc-gateway/BUILD.bazel
@@ -3,7 +3,13 @@ load("@build_stack_rules_proto//rules:proto_plugin.bzl", "proto_plugin")
 proto_plugin(
     name = "protoc-gen-grpc-gateway",
     options = ["paths=source_relative"],
-    tool = "@com_github_grpc_ecosystem_grpc_gateway//protoc-gen-grpc-gateway",
+    tool = "@com_github_grpc_ecosystem_grpc_gateway_v2//protoc-gen-grpc-gateway",
+    visibility = ["//visibility:public"],
+)
+
+proto_plugin(
+    name = "protoc-gen-grpc-openapiv2",
+    tool = "@com_github_grpc_ecosystem_grpc_gateway_v2//protoc-gen-openapiv2",
     visibility = ["//visibility:public"],
 )
 
diff --git rules/ts/proto_ts_library.bzl rules/ts/proto_ts_library.bzl
index ac06a1ef..4087ffb0 100644
--- rules/ts/proto_ts_library.bzl
+++ rules/ts/proto_ts_library.bzl
@@ -26,10 +26,15 @@ def proto_ts_library(
     """
     ts_project(
         name = name,
+        transpiler = "tsc",
         declaration = declaration,
         tsconfig = tsconfig or {
             "compilerOptions": {
+                "moduleResolution": "node",
+                "module": "commonjs",
+                "target": "ESNext",
                 "declaration": declaration,
+                "esModuleInterop": True,
             },
         },
         **kwargs
