diff --git BUILD.bazel BUILD.bazel
index b98441a..0113773 100644
--- BUILD.bazel
+++ BUILD.bazel
@@ -1,10 +1,12 @@
 load("@bazel_skylib//rules:copy_file.bzl", "copy_file")
-
-# Note: if you are still using WORKSPACE, you will need to patch this file to use the following instead
-# load("@python_3_11//:defs.bzl", "py_binary")
-load("@python_versions//3.11:defs.bzl", "py_binary")
+load("@build_stack_rules_proto//rules:proto_compile.bzl", "proto_compile")
+load("@build_stack_rules_proto//rules/cc:proto_cc_library.bzl", "proto_cc_library")
+load("@build_stack_rules_proto//rules/go:proto_go_library.bzl", "proto_go_library")
+load("@build_stack_rules_proto//rules/py:proto_py_library.bzl", "proto_py_library")
+load("@rules_cc//cc:defs.bzl", "cc_library")
 load("@rules_proto//proto:defs.bzl", "proto_library")
 load("@rules_proto_grpc//:defs.bzl", "proto_plugin")
+load("@rules_python//python:defs.bzl", "py_binary")
 load("@rules_python//python:proto.bzl", "py_proto_library")
 load("@rules_python//python/pip_install:requirements.bzl", "compile_pip_requirements")
 load("//extra/bazel:nanopb_cc_proto_library.bzl", "cc_nanopb_proto_library")
@@ -28,6 +30,7 @@ cc_library(
         "pb_decode.h",
         "pb_encode.h",
     ],
+    includes = ["."],
     visibility = ["//visibility:public"],
 )
 
@@ -55,7 +58,9 @@ py_binary(
         "generator",
     ],
     deps = [
-        ":nanopb_py_proto",
+        ":nanopb_py_pb2",
+        "@grpc//src/python/grpcio/grpc:grpcio",
+        "@protobuf//:protobuf_python",
     ],
 )
 
@@ -77,6 +82,20 @@ proto_plugin(
     visibility = ["//visibility:public"],
 )
 
+proto_compile(
+    name = "nanopb_py_compile",
+    outputs = ["nanopb_pb2.py"],
+    plugins = ["@build_stack_rules_proto//plugin/builtin:python"],
+    proto = "nanopb_proto",
+    visibility = ["//visibility:public"],
+)
+
+proto_py_library(
+    name = "nanopb_py_pb2",
+    srcs = ["nanopb_pb2.py"],
+    imports = ["."],
+)
+
 proto_library(
     name = "nanopb_proto",
     srcs = [
@@ -91,6 +110,52 @@ py_proto_library(
     deps = [":nanopb_proto"],
 )
 
+proto_compile(
+    name = "nanopb_cpp_compile",
+    outputs = [
+        "nanopb.pb.cc",
+        "nanopb.pb.h",
+    ],
+    plugins = [
+        "@build_stack_rules_proto//plugin/builtin:cpp",
+        "@clockwork//tools/gazelle:protoc-gen-nolint",
+    ],
+    proto = "nanopb_proto",
+    visibility = ["//visibility:public"],
+)
+
+proto_cc_library(
+    name = "nanopb_cc_library",
+    srcs = ["nanopb.pb.cc"],
+    hdrs = ["nanopb.pb.h"],
+    copts = ["-w"],
+    visibility = ["//visibility:public"],
+    deps = [
+        "@protobuf",
+    ],
+)
+
+proto_compile(
+    name = "nanopb_proto_go_compile",
+    output_mappings = ["nanopb.pb.go=fi.kapsi.koti.jpa/nanopb/nanopb.pb.go"],
+    outputs = ["nanopb.pb.go"],
+    plugins = ["@build_stack_rules_proto//plugin/golang/protobuf:protoc-gen-go"],
+    proto = "nanopb_proto",
+    visibility = ["//visibility:public"],
+)
+
+proto_go_library(
+    name = "nanopb_go_proto",
+    srcs = ["nanopb.pb.go"],
+    importpath = "fi.kapsi.koti.jpa/nanopb",
+    visibility = ["//visibility:public"],
+    deps = [
+        "@org_golang_google_protobuf//reflect/protoreflect",
+        "@org_golang_google_protobuf//runtime/protoimpl",
+        "@org_golang_google_protobuf//types/descriptorpb",
+    ],
+)
+
 cc_nanopb_proto_library(
     name = "test_compilation",
     protos = ["@protobuf//:descriptor_proto"],
diff --git MODULE.bazel MODULE.bazel
index 8aaa68e..6bf8028 100644
--- MODULE.bazel
+++ MODULE.bazel
@@ -1,10 +1,15 @@
 module(
-    name = "nanopb",
+    name = "com_github_nanopb_nanopb",
     version = "0.4.9",
 )
 
 bazel_dep(name = "bazel_skylib", version = "1.7.1")
+bazel_dep(name = "build_stack_rules_proto")
+bazel_dep(name = "clockwork")
+bazel_dep(name = "gazelle")
+bazel_dep(name = "grpc")
 bazel_dep(name = "rules_cc", version = "0.0.10")
+bazel_dep(name = "rules_go")
 bazel_dep(name = "rules_python", version = "0.35.0")
 bazel_dep(name = "rules_proto", version = "6.0.2")
 bazel_dep(name = "protobuf", version = "28.2")
@@ -14,7 +19,7 @@ pip = use_extension("@rules_python//python/extensions:pip.bzl", "pip")
 
 python = use_extension("@rules_python//python/extensions:python.bzl", "python")
 
-PYTHON_VERSION = "3.11"
+PYTHON_VERSION = "3.10"
 
 # No need for this since rules_python already creates our version.
 # python.toolchain(python_version = PYTHON_VERSION)
@@ -24,5 +29,16 @@ use_repo(pip, "nanopb_pypi")
 pip.parse(
     hub_name = "nanopb_pypi",
     python_version = PYTHON_VERSION,
-    requirements_lock = "@nanopb//:extra/requirements_lock.txt",
+    requirements_lock = "@com_github_nanopb_nanopb//:extra/requirements_lock.txt",
 )
+
+go_sdk = use_extension("@rules_go//go:extensions.bzl", "go_sdk")
+go_sdk.from_file(go_mod = "@clockwork//:go.mod")
+
+go_deps = use_extension("@gazelle//:extensions.bzl", "go_deps")
+go_deps.from_file(go_mod = "@clockwork//:go.mod")
+
+use_repo(
+    go_deps,
+    "org_golang_google_protobuf",
+)
\ No newline at end of file
diff --git extra/bazel/nanopb_cc_proto_library.bzl extra/bazel/nanopb_cc_proto_library.bzl
index aae5965..9286edc 100644
--- extra/bazel/nanopb_cc_proto_library.bzl
+++ extra/bazel/nanopb_cc_proto_library.bzl
@@ -39,7 +39,7 @@ cc_nanopb_proto_compile = rule(
         _plugins = attr.label_list(
             providers = [ProtoPluginInfo],
             default = [
-                Label("@nanopb//:nanopb_plugin"),
+                Label("@com_github_nanopb_nanopb//:nanopb_plugin"),
             ],
             doc = "List of protoc plugins to apply",
         ),
@@ -94,5 +94,5 @@ def cc_nanopb_proto_library(name, **kwargs):  # buildifier: disable=function-doc
     )
 
 PROTO_DEPS = [
-    "@nanopb//:nanopb",
+    "@com_github_nanopb_nanopb//:nanopb",
 ]
diff --git extra/bazel/python_deps.bzl extra/bazel/python_deps.bzl
index a66791e..7e03bf1 100644
--- extra/bazel/python_deps.bzl
+++ extra/bazel/python_deps.bzl
@@ -7,6 +7,6 @@ def nanopb_python_deps(interpreter=None):
     if "nanopb_pypi" not in native.existing_rules():
         pip_parse(
             name = "nanopb_pypi",
-            requirements_lock = "@nanopb//:extra/requirements_lock.txt",
+            requirements_lock = "@com_github_nanopb_nanopb//:extra/requirements_lock.txt",
             python_interpreter_target = interpreter,
         )
diff --git generator/nanopb_generator.py generator/nanopb_generator.py
index 16c6fd7..0ede208 100755
--- generator/nanopb_generator.py
+++ generator/nanopb_generator.py
@@ -2039,6 +2039,8 @@ class ProtoFile:
         else:
             yield '/* Generated by %s at %s. */\n\n' % (nanopb_version, time.asctime())
 
+        yield '#pragma GCC system_header\n\n'
+
         if self.fdesc.package:
             symbol = make_identifier(self.fdesc.package + '_' + headername)
         else:
diff --git generator/proto/__init__.py generator/proto/__init__.py
index b2b47b6..dd960b0 100644
--- generator/proto/__init__.py
+++ generator/proto/__init__.py
@@ -71,6 +71,9 @@ def load_nanopb_pb2():
     protosrc = os.path.join(dirname, "nanopb.proto")
     protodst = os.path.join(dirname, "nanopb_pb2.py")
 
+    import nanopb_pb2 as nanopb_pb2_mod
+    return nanopb_pb2_mod
+
     if tmpdir is not None and not os.path.isdir(tmpdir):
         tmpdir = None # Use system-wide temp dir
 
diff --git generator/proto/nanopb.proto generator/proto/nanopb.proto
index 1a1cc98..21892f0 100644
--- generator/proto/nanopb.proto
+++ generator/proto/nanopb.proto
@@ -15,6 +15,7 @@ syntax = "proto2";
 import "google/protobuf/descriptor.proto";
 
 option java_package = "fi.kapsi.koti.jpa.nanopb";
+option go_package = "fi.kapsi.koti.jpa/nanopb";
 
 enum FieldType {
     FT_DEFAULT = 0; // Automatically decide field type, generate static field if possible.
diff --git go.mod go.mod
new file mode 100644
index 0000000..6bfece3
--- /dev/null
+++ go.mod
@@ -0,0 +1,7 @@
+module fi.kapsi.koti.jpa/nanopb
+
+go 1.20
+
+require (
+	google.golang.org/protobuf v1.31.0
+)
diff --git go.sum go.sum
new file mode 100644
index 0000000..2d1cb6c
--- /dev/null
+++ go.sum
@@ -0,0 +1 @@
+google.golang.org/grpc v1.31.0/go.mod h1:N36X2cJ7JwdamYAgDz+s+rVMFjt3numwzf/HckM8pak=
diff --git pb.h pb.h
index 10249bb..6aaeeee 100644
--- pb.h
+++ pb.h
@@ -23,7 +23,7 @@
 /* #define PB_MAX_REQUIRED_FIELDS 256 */
 
 /* Add support for tag numbers > 65536 and fields larger than 65536 bytes. */
-/* #define PB_FIELD_32BIT 1 */
+#define PB_FIELD_32BIT 1
 
 /* Disable support for error messages in order to save some code space. */
 /* #define PB_NO_ERRMSG 1 */
@@ -912,7 +912,7 @@ struct pb_extension_s {
 
 extern "C++"
 {
-namespace nanopb {
+namespace nanopb_msg {
 // Each type will be partially specialized by the generator.
 template <typename GenMessageT> struct MessageDescriptor;
 }  // namespace nanopb
