diff --git BUILD.bazel BUILD.bazel
new file mode 100644
index 0000000..d31c10d
--- /dev/null
+++ BUILD.bazel
@@ -0,0 +1,55 @@
+config_setting(
+    name = "macos",
+    constraint_values = ["@platforms//os:macos"],
+    visibility = ["//visibility:public"],
+)
+
+config_setting(
+    name = "msvc_compiler",
+    flag_values = {"@bazel_tools//tools/cpp:compiler": "msvc-cl"},
+)
+
+cc_library(
+    name = "nanobind",
+    srcs = [
+        "include/nanobind/stl/detail/nb_dict.h",
+        "include/nanobind/stl/detail/nb_list.h",
+        "include/nanobind/stl/detail/traits.h",
+        "src/buffer.h",
+        "src/common.cpp",
+        "src/error.cpp",
+        "src/hash.h",
+        "src/implicit.cpp",
+        "src/nb_enum.cpp",
+        "src/nb_func.cpp",
+        "src/nb_internals.cpp",
+        "src/nb_internals.h",
+        "src/nb_ndarray.cpp",
+        "src/nb_static_property.cpp",
+        "src/nb_type.cpp",
+        "src/trampoline.cpp",
+    ],
+    hdrs = glob(
+        include = [
+            "include/nanobind/*.h",
+            "include/nanobind/eigen/*.h",
+            "include/nanobind/stl/*.h",
+            "include/nanobind/stl/detail/*.h",
+        ],
+        allow_empty = False,
+    ),
+    includes = ["include"],
+    linkopts = select({
+        ":macos": [
+            "-undefined dynamic_lookup",
+            "-Wl,-no_fixup_chains",
+            "-Wl,-dead_strip",
+        ],
+        "//conditions:default": [],
+    }),
+    visibility = ["//visibility:public"],
+    deps = [
+        "@python_3_10//:python_headers",
+        "@robin-map//:robin-map",
+    ],
+)
diff --git MODULE.bazel MODULE.bazel
new file mode 100644
index 0000000..b7ec3be
--- /dev/null
+++ MODULE.bazel
@@ -0,0 +1,15 @@
+module(
+    name = "nanobind",
+    version = "2.1.0",
+    compatibility_level = 1,
+)
+
+bazel_dep(name = "platforms", version = "0.0.10")
+bazel_dep(name = "rules_python", version = "0.36.0")
+bazel_dep(name = "robin-map", version = "1.3.0")
+
+python = use_extension("@rules_python//python/extensions:python.bzl", "python")
+python.toolchain(python_version = "3.10")
+use_repo(python, "python_3_10")
+
+# No need to register the toolchain for Python, we just want the headers.
diff --git include/nanobind/nanobind.h include/nanobind/nanobind.h
index d245a0f..85db9cf 100644
--- include/nanobind/nanobind.h
+++ include/nanobind/nanobind.h
@@ -36,22 +36,22 @@
 #include <new>
 
 // Implementation. The nb_*.h files should only be included through nanobind.h
-#include "nb_python.h"
-#include "nb_defs.h"
-#include "nb_enums.h"
-#include "nb_traits.h"
-#include "nb_tuple.h"
-#include "nb_lib.h"
-#include "nb_descr.h"
-#include "nb_types.h"
-#include "nb_accessor.h"
-#include "nb_error.h"
-#include "nb_attr.h"
-#include "nb_cast.h"
-#include "nb_misc.h"
-#include "nb_call.h"
-#include "nb_func.h"
-#include "nb_class.h"
+#include "nb_python.h"   // IWYU pragma: export
+#include "nb_defs.h"     // IWYU pragma: export
+#include "nb_enums.h"    // IWYU pragma: export
+#include "nb_traits.h"   // IWYU pragma: export
+#include "nb_tuple.h"    // IWYU pragma: export
+#include "nb_lib.h"      // IWYU pragma: export
+#include "nb_descr.h"    // IWYU pragma: export
+#include "nb_types.h"    // IWYU pragma: export
+#include "nb_accessor.h" // IWYU pragma: export
+#include "nb_error.h"    // IWYU pragma: export
+#include "nb_attr.h"     // IWYU pragma: export
+#include "nb_cast.h"     // IWYU pragma: export
+#include "nb_misc.h"     // IWYU pragma: export
+#include "nb_call.h"     // IWYU pragma: export
+#include "nb_func.h"     // IWYU pragma: export
+#include "nb_class.h"    // IWYU pragma: export
 
 #if defined(_MSC_VER)
 #  pragma warning(pop)
diff --git include/nanobind/nb_class.h include/nanobind/nb_class.h
index e1a953f..7954ae7 100644
--- include/nanobind/nb_class.h
+++ include/nanobind/nb_class.h
@@ -480,7 +480,9 @@ public:
     using Base  = typename detail::extract<T, detail::is_base,  Ts...>::type;
     using Alias = typename detail::extract<T, detail::is_alias, Ts...>::type;
 
-    static_assert(sizeof(Alias) < (1 << 24), "Instance size is too big!");
+    // Fix upstream bug. Dscussion here https://github.com/wjakob/nanobind/discussions/759
+    // Upstream fix here will make it to nanobind 2.3: https://github.com/wjakob/nanobind/commit/2aacdcfdd312dabda3d65e683e418e4b3c8c0b04
+    static_assert(sizeof(Alias) < (1ULL << 32), "Instance size is too big!");
     static_assert(alignof(Alias) < (1 << 8), "Instance alignment is too big!");
     static_assert(
         sizeof...(Ts) == !std::is_same_v<Base, T> + !std::is_same_v<Alias, T>,
diff --git include/nanobind/stl/detail/nb_optional.h include/nanobind/stl/detail/nb_optional.h
index 78365c9..deda218 100644
--- include/nanobind/stl/detail/nb_optional.h
+++ include/nanobind/stl/detail/nb_optional.h
@@ -21,9 +21,10 @@ struct optional_caster {
     NB_TYPE_CASTER(Optional, optional_name(Caster::Name))
 
     bool from_python(handle src, uint8_t flags, cleanup_list* cleanup) noexcept {
-        if (src.is_none())
-            // default-constructed value is already empty
+        if (src.is_none()) {
+            value.reset();
             return true;
+        }
 
         Caster caster;
         if (!caster.from_python(src, flags_for_local_caster<T>(flags), cleanup) ||
