diff --git api/BUILD api/BUILD
index bfd3af6..945b020 100644
--- api/BUILD
+++ api/BUILD
@@ -47,6 +47,7 @@ cc_library(
         "@com_google_absl//absl/strings",
         "@com_google_absl//absl/types:variant",
     ],
+    features = ["system_include_paths"],
 )
 
 [config_setting(
diff --git sdk/BUILD sdk/BUILD
index 52ed22a..eda0c6a 100644
--- sdk/BUILD
+++ sdk/BUILD
@@ -9,4 +9,5 @@ cc_library(
     name = "headers",
     hdrs = glob(["include/**/*.h"]),
     strip_include_prefix = "include",
+    features = ["system_include_paths"],
 )
