diff --git BUILD.bazel BUILD.bazel
new file mode 100644
index 0000000..13c3ea3
--- /dev/null
+++ BUILD.bazel
@@ -0,0 +1,7 @@
+
+cc_library(
+    name = "wise_enum",
+    hdrs = glob(["*.h"]),
+    visibility = ["//visibility:public"],
+    includes = ["."],
+)
diff --git MODULE.bazel MODULE.bazel
new file mode 100644
index 0000000..718d1ce
--- /dev/null
+++ MODULE.bazel
@@ -0,0 +1 @@
+module(name = "wise_enum")
diff --git wise_enum_detail.h wise_enum_detail.h
index 1accbe7..2ad8766 100644
--- wise_enum_detail.h
+++ wise_enum_detail.h
@@ -163,7 +163,9 @@ WISE_ENUM_CONSTEXPR_14 bool compare(U u1, U u2) {
 #define WISE_ENUM_IMPL(type, name_storage, friendly, ...)                      \
   WISE_ENUM_IMPL_2(type, WISE_ENUM_IMPL_ONLY_OR_FIRST(name_storage),           \
                    WISE_ENUM_IMPL_STORAGE(name_storage), friendly,             \
-                   WISE_ENUM_IMPL_NARG(__VA_ARGS__), __VA_ARGS__)
+                   WISE_ENUM_IMPL_NARG(__VA_ARGS__), __VA_ARGS__)              \
+  /* Clang sometimes detects this function as unused when no fmtlib calls are present */ \
+  [[maybe_unused]] friendly constexpr auto format_as(WISE_ENUM_IMPL_ONLY_OR_FIRST(name_storage) value) -> decltype(wise_enum::to_string(value)) { return wise_enum::to_string(value); };
 
 #define WISE_ENUM_IMPL_2(type, name, storage, friendly, num_enums, ...)        \
   WISE_ENUM_IMPL_3(type, name, storage, friendly, num_enums,                   \
@@ -182,7 +184,8 @@ WISE_ENUM_CONSTEXPR_14 bool compare(U u1, U u2) {
   namespace detail {                                                           \
   WISE_ENUM_IMPL_ADAPT_2(name, WISE_ENUM_IMPL_NARG(__VA_ARGS__), __VA_ARGS__)  \
   }                                                                            \
-  }
+  }                                                                            \
+  [[maybe_unused]] constexpr auto format_as(name value) -> decltype(wise_enum::to_string(value)) { return wise_enum::to_string(value); };
 
 #define WISE_ENUM_IMPL_ADAPT_2(name, num_enums, ...)                           \
   WISE_ENUM_IMPL_ADAPT_3(name, , num_enums,                                    \
